service/gcs/stdio/stdio.go

```diff
@@ -99,6 +99,136 @@ func (s *ConnectionSet) Files() (_ *FileSet, err error) {
 	return fs, nil
 }
 
+// NewPipeRelay returns a new pipe relay wrapping the given connection stdin, stdout, stderr set.
+func (s *ConnectionSet) NewPipeRelay() (_ *PipeRelay, err error) {
+	pr := &PipeRelay{s: s}
+	defer func() {
+		if err != nil {
+			pr.closePipes()
+		}
+	}()
+
+	if s.In != nil {
+		pr.pipes[0], pr.pipes[1], err = os.Pipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdin pipe relay")
+		}
+	}
+	if s.Out != nil {
+		pr.pipes[2], pr.pipes[3], err = os.Pipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdout pipe relay")
+		}
+	}
+	if s.Err != nil {
+		pr.pipes[4], pr.pipes[5], err = os.Pipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stderr pipe relay")
+		}
+	}
+	return pr, nil
+}
+
+// PipeRelay is a relay built to expose a pipe interface
+// for stdin, stdout, stderr ontop of a ConnectionSet.
+type PipeRelay struct {
+	wg sync.WaitGroup
+	s  *ConnectionSet
+	// pipes format is stdin [0 read, 1 write], stdout [2 read, 3 write], stderr [4 read, 5 write].
+	pipes [6]*os.File
+}
+
+// Files returns a FileSet with an os.File for each connection
+// in the connection set.
+func (pr *PipeRelay) Files() (*FileSet, error) {
+	fs := new(FileSet)
+	if pr.s.In != nil {
+		fs.In = pr.pipes[0]
+	}
+	if pr.s.Out != nil {
+		fs.Out = pr.pipes[3]
+	}
+	if pr.s.Err != nil {
+		fs.Err = pr.pipes[5]
+	}
+	return fs, nil
+}
+
+// Start starts the relay operation. The caller must call Wait to wait
+// for the relay to finish and release the associated resources.
+func (pr *PipeRelay) Start() {
+	if pr.s.In != nil {
+		pr.wg.Add(1)
+		go func() {
+			if _, err := io.Copy(pr.pipes[1], pr.s.In); err != nil {
+				logrus.Errorf("error copying stdin to pipe: %s", err)
+			}
+			if err := pr.pipes[1].Close(); err != nil {
```

Comment:
The EOF needs to be set to the child process. So closing the `write` side does this. The outstanding call to `Wait()` will eventually close both sides and the sockets when out/err complete too so its fine here.